src/compiler/crystal/codegen/primitives.cr

```diff
@@ -260,20 +260,62 @@ class Crystal::CodeGenVisitor
   end
 
   private def codegen_out_of_range(target_type : IntegerType, arg_type : FloatType, arg)
+    # we allow one comparison to be unordered so that NaNs are caught
     if arg_type.kind == :f32 && target_type.kind == :u128
-      # since Float32::MAX < UInt128::MAX
-      # the range checking is replaced by a positive check only
-      builder.fcmp(LLVM::RealPredicate::OLT, arg, llvm_type(arg_type).const_float(0))
+      # since `Float32::MAX < UInt128::MAX`, we cannot make the upper bound
+      # equal to `UInt128::MAX.to_f32` there is still an overflow if
+      # `arg` is positive infinity
+      or(
+        builder.fcmp(LLVM::RealPredicate::ULT, arg, float(0, arg_type)),
+        builder.fcmp(LLVM::RealPredicate::OGT, arg, float(Float32::MAX, arg_type))
+      )
     else
       min_value, max_value = target_type.range
-      # arg < min_value || arg > max_value
+      if arg_type.bytes <= target_type.bytes
+        # if the float type has fewer bits of precision than the integer type
+        # (although we use the number of bytes instead the results are the same)
+        # then the upper bound would mistakenly allow values near the upper
+        # limit, e.g. 2147483647_i32 -> 2147483648_f32, because the bound itself
+        # is rounded to the nearest even-significand number; we choose the
+        # predecessor as the upper bound, i.e. 2147483520_f32, ensuring it is
+        # exact when converted back to an integer
+        max_value = arg_type.kind == :f32 ? float32_upper_bound(max_value) : float64_upper_bound(max_value)
+      end
+
+      # !(arg >= min_value) || arg > max_value
       or(
-        builder.fcmp(LLVM::RealPredicate::OLT, arg, int_to_float(target_type, arg_type, int(min_value, target_type))),
+        builder.fcmp(LLVM::RealPredicate::ULT, arg, int_to_float(target_type, arg_type, int(min_value, target_type))),
         builder.fcmp(LLVM::RealPredicate::OGT, arg, int_to_float(target_type, arg_type, int(max_value, target_type)))
       )
     end
   end
 
+  private def float32_upper_bound(max_value)
+    {% begin %}
+      case max_value
+      when Int32, UInt32, Int64, UInt64 {% unless flag?(:windows) %}, Int128, UInt128 {% end %}
```

Comment:
```suggestion
      when Int32, UInt32, Int64, UInt64 {% unless flag?(:win32) %}, Int128, UInt128 {% end %}
```

The symbols are missing when targeting `win32` API / msvc. It's not a `windows` platform issue.

Reply:
Would `flag?(:msvc)` be more appropriate here?

Reply:
Maybe. But the established flag is `win32`.